qdglue/metrics/grid_based_metrics.py

```diff
@@ -0,0 +1,290 @@
+
+        features_indexes = np.ravel_multi_index(
+            multi_index=features_multi_indexes.T, dims=self.resolution, mode="raise"
+        )  # TODO check this line
+
+        # Collecting the best fitnesses per feature index
+        best_fitness_per_feature_index: Dict[int, Fitness] = dict()
+
+        for index_feature, fitness in zip(features_indexes, fitnesses):
+            if index_feature not in best_fitness_per_feature_index:
+                best_fitness_per_feature_index[index_feature] = fitness
+            else:
+                best_fitness_per_feature_index[index_feature] = np.maximum(
+                    best_fitness_per_feature_index[index_feature], fitness
+                )
+
+        best_fitnesses = np.asarray(list(best_fitness_per_feature_index.values()))
+
+        return best_fitnesses
+
+    @property
+    def num_cells(self) -> int:
+        return np.prod(self.resolution).astype(np.int32).item()
+
+    def _check_validity_features(self, features: Feature) -> None:
+        assert np.all(np.min(features, axis=0) >= self.feature_space.low) and np.all(
+            np.max(features, axis=0) <= self.feature_space.high
+        ), "Features must be within the feature space bounds"
+
+
+class CVTMetricsCalculator(DiscreteArchiveMetricsCalculator):
+    def __init__(
+        self,
+        fitness_bounds: gymnasium.spaces.Box,
+        centroids: Feature,
+        num_points_ccdf: int = 100,
+    ):
+        super().__init__(fitness_bounds, num_points_ccdf)
+
+        self.centroids = centroids
+
+    @property
+    def num_cells(self) -> int:
+        return self.centroids.shape[0]
+
+    def _check_validity_features(self, features: Feature) -> None:
+        pass
+
+    def _get_index_closest_centroid(self, feature: Feature) -> int:
+        distances = np.linalg.norm(self.centroids - feature, axis=1)
+        return np.argmin(distances).item()
+
+    def _extract_best_fitnesses_per_cell(
+        self,
+        fitnesses: Fitness,
+        features: Feature,
+    ) -> Fitness:
+        """
+        Returns an array of the best fitnesses per cell.
+        The empty cells do not appear in the returned array.
+        """
+
+        best_fitness_per_centroid_index: Dict[int, Fitness] = dict()
+
+        for feature, fitness in zip(features, fitnesses):
+            index_closest_centroid = self._get_index_closest_centroid(feature)
+            if index_closest_centroid not in best_fitness_per_centroid_index:
+                best_fitness_per_centroid_index[index_closest_centroid] = fitness
+            else:
+                best_fitness_per_centroid_index[index_closest_centroid] = np.maximum(
+                    best_fitness_per_centroid_index[index_closest_centroid], fitness
+                )
+
+        best_fitnesses = np.asarray(list(best_fitness_per_centroid_index.values()))
+
+        return best_fitnesses
+
+
+def example_grid_based_metrics():
```

Comment:
Move this to an example file or (even better) a test?

Reply:
moved to a test :)